services/src/projects/projectdb.rs

```diff
@@ -0,0 +1,383 @@
+use crate::projects::project::{OrderBy, ProjectFilter, ProjectListing, ProjectOwner, ProjectId, Project, CreateProject, UpdateProject, UserProjectPermission, ProjectPermission, ProjectListOptions};
+use crate::users::user::{UserIdentification, Validated};
+use std::collections::HashMap;
+use crate::error::Result;
+use crate::error;
+use snafu::ensure;
+
+pub trait ProjectDB: Send + Sync {
+    // TODO: pass session instead of UserIdentification?
+    fn list(&self, user: UserIdentification, options: Validated<ProjectListOptions>) -> Vec<ProjectListing>;
+    fn load(&self, user: UserIdentification, project: ProjectId) -> Result<Project>;
+    fn create(&mut self, user: UserIdentification, project: Validated<CreateProject>) -> ProjectId;
+    fn update(&mut self, user: UserIdentification, project: Validated<UpdateProject>) -> Result<()>;
+    fn delete(&mut self, user: UserIdentification, project: ProjectId) -> Result<()>;
+
+    fn list_permissions(&mut self, user: UserIdentification, project: ProjectId) -> Result<Vec<UserProjectPermission>>;
+    fn add_permission(&mut self, user: UserIdentification, permission: UserProjectPermission) -> Result<()>;
+    fn remove_permission(&mut self, user: UserIdentification, permission: UserProjectPermission) -> Result<()>;
+}
+
+#[derive(Default)]
+pub struct HashMapProjectDB {
+    projects: HashMap<ProjectId, Project>,
+    permissions: Vec<UserProjectPermission>,
+}
+
+// TODO: versioning?
+impl ProjectDB for HashMapProjectDB {
```

Comment:
You should split the ProjectDB trait from the implementations.

Reply:
split how? different files? why?

Reply:
I was thinking of one for the traits and one for each implementation.